workspaces/adventure-pack/goodies/typescript/Iterator.prototype.product/index.ts

```diff
@@ -0,0 +1,21 @@
+import "../Iterator.prototype.toIterable";
+import { iteratorPrototype } from "../Iterator.prototype";
+
+declare global {
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  interface Iterator<T> {
+    product(this: Iterator<number>): number;
+  }
+}
+
+iteratorPrototype.product = function (this: Iterator<number>): number {
+  let res = 1;
+  let isEmpty = true;
+
+  for (const element of this.toIterable()) {
+    isEmpty = false;
+    res *= element;
+  }
+
+  return isEmpty ? 0 : res;
```

Comment:
Mathematically, it's actually reasonable to return 1 for an empty product, similar to how zero factorial is defined to be 1, thus making many combinatorics formulas continue to work even for zero.

So we don't need to special-case the empty case.